src/Microsoft.ML.Data/MLContext.cs

```diff
@@ -104,12 +108,27 @@ private void ProcessMessage(IMessageSource source, ChannelMessage message)
             log(this, new LoggingEventArgs(msg));
         }
 
-        bool IHostEnvironment.IsCancelled => _env.IsCancelled;
         string IExceptionContext.ContextDescription => _env.ContextDescription;
         TException IExceptionContext.Process<TException>(TException ex) => _env.Process(ex);
-        IHost IHostEnvironment.Register(string name, int? seed, bool? verbose) => _env.Register(name, seed, verbose);
+        IHost IHostEnvironment.Register(string name, int? seed, bool? verbose)
+        {
+            var host = _env.Register(name, seed, verbose);
```

Comment:
Does this Register actually get called? Since we are passing the inner `_env` to all of the Catalogs above, that is the object that all the code is calling `Register` on right? There is a split here between `MLContext` and `HostEnivronmentBase`.

Reply:
Yes, it gets called. From an environment you can create multiple hosts and then from each host you can create multiple hosts. I have verified this with the test that I wrote.

Reply:
Your test calls Register directly on the MLContext, which is not what users are going to do.

Users are going to write code like:

```C#
var mlContext = new MLContext();

mlContext.Regression.Trainers.Sdca();
```

If you walk that code path above - `mlContext.Regression` is:

https://github.com/dotnet/machinelearning/blob/b6f94bcb86b4fa25aa0efed2ecc036ae981c98e0/src/Microsoft.ML.Data/MLContext.cs#L86

Which gets the underlying `_env` passed into it. NOT the `MLContext`. Therefore, when Sdca calls `Register` it will call it on the inner `_env`, not on the `MLContext`.

Reply:
I expect the user to cancel via MLContext. I will sync offline with on you on this.

Reply:
Done.